apps/passport-server/src/services/zuconnectTripshaSyncService.ts

```diff
@@ -0,0 +1,196 @@
+import { SemaphoreService } from "./semaphoreService";
+import { setError, traced } from "./telemetryService";
+
+const NAME = "Zuconnect Tripsha";
+
+/**
+ * Fetches ticket data from Tripsha's API and stores it in the database.
+ */
+export class ZuconnectTripshaSyncService {
+  private zuconnectTripshaAPI: IZuconnectTripshaAPI;
+  private rollbarService: RollbarService | null;
+  private semaphoreService: SemaphoreService;
+  private context: ApplicationContext;
+  private timeout: NodeJS.Timeout | undefined;
+  private static readonly SYNC_INTERVAL_MS = 1000 * 60;
+
+  public constructor(
+    context: ApplicationContext,
+    api: IZuconnectTripshaAPI,
+    rollbarService: RollbarService | null,
+    semaphoreService: SemaphoreService
+  ) {
+    this.context = context;
+    this.zuconnectTripshaAPI = api;
+    this.rollbarService = rollbarService;
+    this.semaphoreService = semaphoreService;
+  }
+
+  /**
+   * Starts the sync service by conducting an initial sync and setting up
+   * a timeout to repeat it at a set interval.
+   */
+  public async start(): Promise<void> {
+    logger("[ZUCONNECT TRIPSHA] Starting sync loop");
+
+    const trySync = async (): Promise<void> => {
+      await this.trySync();
+      this.timeout = setTimeout(
+        () => trySync(),
+        ZuconnectTripshaSyncService.SYNC_INTERVAL_MS
+      );
+    };
+
+    trySync();
+  }
+
+  /**
+   * Stop the sync service by cancelling the timeout.
+   */
+  public stop(): void {
+    if (this.timeout) {
+      clearTimeout(this.timeout);
+    }
+  }
+
+  /**
+   * Run a sync job and report any errors that are thrown.
+   */
+  public async trySync(): Promise<void> {
+    return traced(NAME, "trySync", async (span) => {
+      try {
+        logger("[ZUCONNECT TRIPSHA] Sync start");
+        await this.sync();
+        this.semaphoreService.scheduleReload();
+        logger("[ZUCONNECT TRIPSHA] Sync finished");
+      } catch (e) {
+        this.rollbarService?.reportError(e);
+        logger("[ZUCONNECT TRIPSHA] Sync failed", e);
+        setError(e, span);
+      }
+    });
+  }
+
+  /**
+   * Fetch and save data from Tripsha. Thrown errors will be caught in trySync.
+   */
+  public async sync(): Promise<void> {
+    return traced(NAME, "sync", async (_span) => {
+      const tickets = await this.fetchData();
+      await this.saveData(tickets);
+    });
+  }
+
+  /**
+   * Fetch tickets from the Tripsha API. Will throw errors if the network
+   * fails or if invalid data is received.
+   */
+  private async fetchData(): Promise<ZuconnectTicket[]> {
+    return this.zuconnectTripshaAPI.fetchTickets();
+  }
+
+  /**
+   * Convert a Tripsha ticket type to a product ID.
+   */
+  private ticketTypeToProductId(type: ZuconnectTicket["type"]): string {
+    return ZUCONNECT_PRODUCT_ID_MAPPINGS[type].id;
+  }
+
+  private isMockTicketRecord(ticket: ZuconnectTicketDB): boolean {
+    return ticket.external_ticket_id.startsWith("mock-");
+  }
```

Comment:
this is another implicit thing to remember - maybe worth adding a column to the tripsha tickets table specifying that it's a mocket ticket. up to you.